ReactNativeClient/lib/services/RecordAudio.js

```diff
@@ -0,0 +1,96 @@
+import AudioRecorderPlayer from 'react-native-audio-recorder-player';
+import { PermissionsAndroid, Alert, Platform } from 'react-native';
+
+class RecordAudio {
+	constructor() {
+		this.dispatch = () => {};
+		this.audioRecorderPlayer = new AudioRecorderPlayer();
+		this.recordSecs = 0;
+		this.recordTime = 0;
+	}
+
+	static instance() {
+		if (RecordAudio.instance_) return RecordAudio.instance_;
+		RecordAudio.instance_ = new RecordAudio();
+		return RecordAudio.instance_;
+	}
+
+	async onStartRecord() {
+		if (Platform.OS === 'android') {
+			try {
+				const granted = await PermissionsAndroid.request(
+					PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
+					{
+						title: 'Permissions for write access',
+						message: 'Give permission to your storage to write a file',
+						buttonPositive: 'ok',
+					},
+				);
+				if (granted === PermissionsAndroid.RESULTS.GRANTED) {
+					console.log('You can use the storage');
+				} else {
+					console.log('permission denied');
+					return;
+				}
+			} catch (err) {
+				console.warn(err);
+				Alert.alert('Permission Needed', 'Unable to get storage permission.');
+				return;
+			}
+		}
+		if (Platform.OS === 'android') {
+			try {
+				const granted = await PermissionsAndroid.request(
+					PermissionsAndroid.PERMISSIONS.RECORD_AUDIO,
+					{
+						title: 'Permissions for recording audio',
+						message: 'Give permission to record audio from the mic',
+						buttonPositive: 'ok',
+					},
+				);
+				if (granted === PermissionsAndroid.RESULTS.GRANTED) {
+					console.log('You can record audio');
+				} else {
+					console.log('permission denied');
+					return;
+				}
+			} catch (err) {
+				console.warn(err);
+				Alert.alert('Permission Needed', 'Unable to get microphone permission.');
+				return;
+			}
```

Comment:
I'm not so keen on big nested if/else statements like this one. Could you perhaps refactor into a function as is done in screens/config.js. There's also the issue mentioned by @roman-r-m that PermissionAndroid doesn't work well but we can leave it for now. If it's separated in a separate function it will be easier to change the permission package later on.

Reply:
Good point, thanks. I actually wanted to extract this since I used the same code in another PR here: https://github.com/laurent22/joplin/pull/3202